packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js

```diff
@@ -88,175 +93,109 @@ function widgetToBlock( widget ) {
+					prevBlocks.map( ( block ) => [
+						getWidgetId( block ),
+						block,
+					] )
 				);
 
-				const nextBlocks = _nextBlocks.map( ( nextBlock, index ) => {
-					if (
-						nextBlock.attributes.__internalWidgetId &&
-						nextBlock.attributes.__internalWidgetId in
-							blocksByWidgetId
-					) {
-						const block =
-							blocksByWidgetId[
-								nextBlock.attributes.__internalWidgetId
-							];
-						if ( ! isEqual( block, nextBlock ) ) {
-							const widgetId =
-								nextBlock.attributes.__internalWidgetId;
-							const widgetToUpdate = sidebar.getWidget(
-								widgetId
-							);
-							const widget = blockToWidget(
-								nextBlock,
-								widgetToUpdate
-							);
-							sidebar.updateWidget( widget );
+				const nextWidgets = nextBlocks.map( ( nextBlock ) => {
+					const widgetId = getWidgetId( nextBlock );
+
+					// Update existing widgets.
+					if ( widgetId && prevBlocksMap.has( widgetId ) ) {
+						const prevBlock = prevBlocksMap.get( widgetId );
+						const prevWidget = sidebar.getWidget( widgetId );
+
+						// Bail out updates by returning the previous widgets.
+						// Deep equality is necessary until the block editor's internals changes.
+						if ( isEqual( nextBlock, prevBlock ) ) {
+							return prevWidget;
 						}
-						return nextBlock;
+
+						return blockToWidget( nextBlock, prevWidget );
 					}
 
-					const widget = blockToWidget( nextBlock );
-					const widgetId = sidebar.addWidget( widget, index );
-					return {
-						...nextBlock,
-						attributes: {
-							...nextBlock.attributes,
-							__internalWidgetId: widgetId,
-						},
-					};
+					// Add a new widget.
+					return blockToWidget( nextBlock );
 				} );
 
-				const seen = nextBlocks.map(
-					( block ) => block.attributes.__internalWidgetId
-				);
-
-				for ( const block of lastState ) {
-					const widgetId = block.attributes.__internalWidgetId;
-					if ( ! seen.includes( widgetId ) ) {
-						sidebar.removeWidget( widgetId );
-					}
-				}
+				const addedWidgetIds = sidebar.setWidgets( nextWidgets );
 
 				if (
-					nextBlocks.length === lastState.length &&
-					! isEqual(
-						nextBlocks.map(
-							( { attributes: { __internalWidgetId } } ) =>
-								__internalWidgetId
-						),
-						lastState.map(
-							( { attributes: { __internalWidgetId } } ) =>
-								__internalWidgetId
-						)
-					)
+					addedWidgetIds.filter( ( widgetId ) => widgetId !== null )
+						.length
 				) {
-					const order = nextBlocks.map(
-						( { attributes: { __internalWidgetId } } ) =>
-							__internalWidgetId
-					);
-					sidebar.setWidgetIds( order );
-				}
+					return nextBlocks.map( ( nextBlock, index ) => {
+						const addedWidgetId = addedWidgetIds[ index ];
+
+						if ( addedWidgetId !== null ) {
+							return addWidgetIdToBlock(
+								nextBlock,
+								addedWidgetId
+							);
+						}
 
-				ignoreIncoming.current = false;
+						return nextBlock;
+					} );
+				}
```

Comment:
We can combine the `filter` of `addedWidgetIds` and the `map` of `nextBlocks` here if we use `reduce`. That way there's only one array traversal. I'm not sure if it would actually make a difference to real world performance, though.

```js
const addedWidgetIds = sidebar.setWidgets( nextWidgets );

return nextBlocks.reduce(
	( updatedNextBlocks, nextBlock, index ) => {
		const addedWidgetId = addedWidgetIds[ index ];

		if ( addedWidgetId !== null ) {
			if ( updatedNextBlocks === nextBlocks ) {
				updatedNextBlocks = { ...nextBlocks };
			}
			updatedNextBlocks[ index ] = addWidgetIdToBlock(
				nextBlock,
				addedWidgetId
			);
		}

		return updatedNextBlocks;
	},
	nextBlocks
);
```

Reply:
I don't think there will be any difference since it's not the hot path and it's still O(N), but I like how it looks so I'll steal it 😆.